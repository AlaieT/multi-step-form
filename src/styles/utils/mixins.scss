@mixin flex($param: null, $type: null) {
  display: flex;

  @if  $param == "center&center" {
    justify-content: center;
    align-items: center;
  }

  @if  $param == "center&end" {
    justify-content: center;
    align-items: flex-end;
  }

  @if  $param == "end&center" {
    justify-content: flex-end;
    align-items: center;
  }

  @if  $param == "space-between&center" {
    justify-content: space-between;
    align-items: center;
  }

  @if  $type == "col" {
    flex-direction: column;
  }
}
