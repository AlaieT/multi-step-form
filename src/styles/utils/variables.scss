$background-white: rgb(255 255 255);
$background-global: rgb(235 243 254);
$background-select-step: rgb(191 226 253);
$primary-text-color: rgb(2 41 90);
$secondary-text-color: rgb(150 153 171);
$plan-border-color: rgb(85 75 209);
$summary-hover-button: rgb(133 127 250);
$select-input: rgb(71 61 255);
$select-button: rgb(0 62 138);
$select-plan: rgb(246 247 254);
$error: hsl(354deg 84% 57%);
$shadow-color: rgb(214 217 230);
$content-box-shadow: 0 25px 40px -20px rgba(0 0 0 / 10%);
