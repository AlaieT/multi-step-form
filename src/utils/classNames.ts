const classNames = (...classes: (string | undefined)[]): string | undefined => {
  return classes.filter((item) => !!item).join(" ");
};

export default classNames;
