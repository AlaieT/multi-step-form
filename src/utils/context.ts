import { createContext } from "react";

import type { FormContextType } from "../types";

const FormContext = createContext<FormContextType>({
  form: {
    planMode: "yearly",
    priceMode: "yr"
  },
  setForm: () => undefined
});

export { FormContext };
