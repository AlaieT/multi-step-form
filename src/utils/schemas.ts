import { object, string, boolean, number } from "yup";

const multiStepFromSchema = object({
  name: string().required("Name is a required field!"),
  email: string()
    .required("Email is a required field!")
    .email("Must be a valid email!"),
  phone: string().required("Phone is a required field!"),
  plan: string().oneOf(["arcade", "advanced", "pro"]).defined(),
  planMode: boolean().defined(),
  onlineService: boolean().defined(),
  largeStorage: boolean().defined(),
  customizableProfile: boolean().defined(),
  totalPrice: number().defined(),
});

export { multiStepFromSchema };
