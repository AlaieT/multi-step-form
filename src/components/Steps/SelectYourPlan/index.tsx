import React, { useContext, useState } from "react";

import ToggleSwitch from "../../ToggleSwitch";
import Button from "../../Button";
import Wrap from "../Wrap";

import type { SelectYourPlanProps } from "../../../types";

import { ReactComponent as IconArcade } from "../../../assets/images/icon-arcade.svg";
import { ReactComponent as IconAdvanced } from "../../../assets/images/icon-advanced.svg";
import { ReactComponent as IconPro } from "../../../assets/images/icon-pro.svg";

import styles from "../../../styles/components/steps/selectYourPlan.module.scss";
import { FormContext } from "../../../utils/context";

const SelectYourPlan = ({
  plans: { arcade, advanced, pro },
  register,
  getValues,
  changeStep
}: SelectYourPlanProps) => {
  const { form, setForm } = useContext(FormContext);
  const [planMode, setPlanMode] = useState(getValues("planMode"));

  return (
    <Wrap>
      <div id={styles.content}>
        <div id={styles.header}>
          <h1>Select your plan</h1>
          <h2>You have the option of monthly or yearly billing.</h2>
        </div>
        <div className={styles.option}>
          <input
            {...register("plan")}
            id="arcade"
            type="radio"
            value="arcade"
            defaultChecked={getValues("plan") === "arcade"}
          />
          <IconArcade width={40} height={40} />
          <label htmlFor="arcade">Arcade</label>
          <p>{`$${arcade[form.planMode]}/${form.priceMode}`}</p>
          {planMode && <p>2 months free</p>}
        </div>
        <div className={styles.option}>
          <input
            {...register("plan")}
            id="advanced"
            type="radio"
            value="advanced"
            defaultChecked={getValues("plan") === "advanced"}
          />
          <IconAdvanced width={40} height={40} />
          <label htmlFor="advanced">Advanced</label>
          <p>{`$${advanced[form.planMode]}/${form.priceMode}`}</p>
          {planMode && <p>2 months free</p>}
        </div>
        <div className={styles.option}>
          <input
            {...register("plan")}
            id="pro"
            type="radio"
            value="pro"
            defaultChecked={getValues("plan") === "pro"}
          />
          <IconPro width={40} height={40} />
          <label htmlFor="pro">Pro</label>
          <p>{`$${pro[form.planMode]}/${form.priceMode}`}</p>
          {planMode && <p>2 months free</p>}
        </div>
        <div id={styles.switch}>
          <label id={!planMode ? styles.selected : undefined} htmlFor="switch">
            Monthly
          </label>
          <ToggleSwitch
            {...register("planMode")}
            id="switch"
            onClick={() => {
              setPlanMode((item) => !item);
              setForm({
                planMode: planMode ? "monthly" : "yearly",
                priceMode: planMode ? "mo" : "yr"
              });
            }}
            defaultChecked={getValues("planMode")}
          />
          <label id={planMode ? styles.selected : undefined} htmlFor="switch">
            Yearly
          </label>
        </div>
      </div>
      <div id={styles.controls}>
        <Button id={styles.go_back} type="button" onClick={() => changeStep(0)}>
          Go Back
        </Button>
        <Button
          id={styles.next_step}
          type="button"
          onClick={() => changeStep(2)}
        >
          Next Step
        </Button>
      </div>
    </Wrap>
  );
};

export default SelectYourPlan;
