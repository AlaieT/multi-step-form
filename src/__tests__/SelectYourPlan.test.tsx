import React from "react";
import {
  render,
  cleanup,
  screen,
  fireEvent,
  renderHook,
  waitFor
} from "@testing-library/react";
import { useForm } from "react-hook-form";
import { yupResolver } from "@hookform/resolvers/yup";

import SelectYourPlan from "../components/Steps/SelectYourPlan";
import { multiStepFromSchema } from "../utils/schemas";
import { act } from "react-dom/test-utils";

afterEach(cleanup);

describe("SelectYourPlan", () => {
  const props = {
    plans: {
      arcade: {
        monthly: 9,
        yearly: 90
      },
      advanced: {
        monthly: 12,
        yearly: 120
      },
      pro: {
        monthly: 15,
        yearly: 150
      }
    }
  };
  const useFormProps = {
    resolver: yupResolver(multiStepFromSchema),
    defaultValues: {
      name: "",
      email: "",
      phone: "",
      plan: "arcade",
      planMode: true,
      customizableProfile: false,
      onlineService: false,
      largeStorage: false,
      totalPrice: 0
    },
    mode: "onChange"
  };
  const changeStepMock = jest.fn();

  describe("correct render", () => {
    it("should match snapshot", () => {
      const { result } = renderHook(() => useForm(useFormProps));
      expect(
        render(
          <SelectYourPlan
            register={result.current.register}
            getValues={result.current.getValues}
            plans={props.plans}
            changeStep={changeStepMock}
          />
        ).asFragment()
      ).toMatchSnapshot();
    });
  });

  describe("correct functionality", () => {
    it("should correct select any plan", async () => {
      const { result } = renderHook(() => useForm(useFormProps));

      render(
        <SelectYourPlan
          register={result.current.register}
          getValues={result.current.getValues}
          plans={props.plans}
          changeStep={changeStepMock}
        />
      );

      const [arcade, advanced, pro] = screen.getAllByRole("radio");

      expect(arcade).toBeChecked();
      expect(advanced).not.toBeChecked();
      expect(pro).not.toBeChecked();
      expect(result.current.getValues("plan")).toBe("arcade");

      fireEvent.click(advanced);

      expect(arcade).not.toBeChecked();
      expect(advanced).toBeChecked();
      expect(pro).not.toBeChecked();
      expect(result.current.getValues("plan")).toBe("advanced");

      fireEvent.click(pro);

      expect(arcade).not.toBeChecked();
      expect(advanced).not.toBeChecked();
      expect(pro).toBeChecked();
      expect(result.current.getValues("plan")).toBe("pro");
    });

    it("should switch payment mode on switch", async () => {
      const { result } = renderHook(() => useForm(useFormProps));

      const { rerender } = render(
        <SelectYourPlan
          register={result.current.register}
          getValues={result.current.getValues}
          plans={props.plans}
          changeStep={changeStepMock}
        />
      );

      expect(screen.getAllByText("2 months free")).toHaveLength(3);

      await act(async () => {
        fireEvent.click(screen.getByRole("checkbox"));
      });

      expect(screen.queryByText("2 months free")).toBeNull();
    });
  });
});
