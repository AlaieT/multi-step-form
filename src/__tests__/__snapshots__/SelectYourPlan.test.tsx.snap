// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectYourPlan correct render should match snapshot 1`] = `
<DocumentFragment>
  <div
    id="step"
  >
    <div
      id="content"
    >
      <div
        id="header"
      >
        <h1>
          Select your plan
        </h1>
        <h2>
          You have the option of monthly or yearly billing.
        </h2>
      </div>
      <div
        class="option"
      >
        <input
          checked=""
          id="arcade"
          name="plan"
          type="radio"
          value="arcade"
        />
        <div
          height="40"
          width="40"
        />
        <label
          for="arcade"
        >
          Arcade
        </label>
        <p>
          $90/yr
        </p>
        <p>
          2 months free
        </p>
      </div>
      <div
        class="option"
      >
        <input
          id="advanced"
          name="plan"
          type="radio"
          value="advanced"
        />
        <div
          height="40"
          width="40"
        />
        <label
          for="advanced"
        >
          Advanced
        </label>
        <p>
          $120/yr
        </p>
        <p>
          2 months free
        </p>
      </div>
      <div
        class="option"
      >
        <input
          id="pro"
          name="plan"
          type="radio"
          value="pro"
        />
        <div
          height="40"
          width="40"
        />
        <label
          for="pro"
        >
          Pro
        </label>
        <p>
          $150/yr
        </p>
        <p>
          2 months free
        </p>
      </div>
      <div
        id="switch"
      >
        <label
          for="switch"
        >
          Monthly
        </label>
        <div
          class="switch"
        >
          <input
            checked=""
            id="switch"
            name="planMode"
            type="checkbox"
          />
        </div>
        <label
          for="switch"
          id="selected"
        >
          Yearly
        </label>
      </div>
    </div>
    <div
      id="controls"
    >
      <button
        class="button"
        id="go_back"
        type="button"
      >
        Go Back
      </button>
      <button
        class="button"
        id="next_step"
        type="button"
      >
        Next Step
      </button>
    </div>
  </div>
</DocumentFragment>
`;
